File: src/commands/group/removeadmin.ts
import {Command, flags as flagtype} from "@oclif/command";
import chalk from "chalk";
import {GroupUserEditResponse, SDKError, ErrorCodes} from "@ironcorelabs/ironnode";
import {ironnode} from "../../lib/SDK";
import {userList, keyFile} from "../../lib/sharedFlags";
import * as GroupMaps from "../../lib/GroupMaps";
import {createDisplayTable, buildCommandSampleText} from "../../lib/Utils";

/**
 * Group remove admin command. Takes a comma-separated list of emails as well as the group name and attempts to remove users as admins from the group.
 */
export default class RemoveAdmin extends Command {
    static aliases = ["group:removeadmins", "group:removeAdmin", "group:removeAdmins"];
    static description = "Remove admins from a group given their email address.";
    static args = [
        {
            name: "group",
            description: "Name of the group. Can alternately refer to a group by ID. Indicate IDs by prefixing with 'id^' e.g. 'id^groupID'.",
            required: true,
        },
    ];
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
        users: userList("Remove admin permissions from the comma-separated list of user emails.")(),
    };
    static examples = [buildCommandSampleText("group:removeadmin -u test@example.com,test2@example.com myGroup")];

    /**
     * Build up a table that displays the results from the group remove admin operation showing both successes and failures.
     */
    buildResultTable(buildResults: GroupUserEditResponse) {
        const table = createDisplayTable(["User", "Result"]);

        buildResults.succeeded.map((removedUser) => {
            table.push([removedUser, chalk.green("✔ Removed as admin")]);
        });

        buildResults.failed.map((removedUser) => {
            table.push([removedUser.id, chalk.red(removedUser.error)]);
        });

        return table;
    }

    async run() {
        const {args, flags} = this.parse(RemoveAdmin);
        const groupID = await GroupMaps.getGroupIDFromName(args.group);

        ironnode()
            .group.removeAdmins(groupID, flags.users as string[])
            .then((removeResult) => {
                this.log(`\n${this.buildResultTable(removeResult).toString()}\n`);
            })
            .catch((e: SDKError) => {
                if (e.code === ErrorCodes.GROUP_REMOVE_ADMINS_REQUEST_FAILURE) {
                    this.log(e.message);
                    return this.error(chalk.red(`Unable to complete request to remove admins from the group.`));
                }
                this.error(chalk.red(e.message));
            });
    }
}

File: src/group/remove_admin.rs
use std::path::PathBuf;

use clap::Parser;
use ironoxide::{
    IronOxideErr,
    group::GroupAccessEditResult,
    prelude::{BlockingIronOxide, GroupId, GroupName, UserId},
};
use itertools::Either;
use yansi::Paint;

use crate::{
    group_maps::{convert_group_names_to_ids, get_group_maps},
    util,
};

const EXAMPLE: &str = "EXAMPLE

    $ ironhide group remove-admin -u test@example.com,test2@example.com myGroup

";

#[derive(Parser)]
/// Remove admins from a group given their email address.
#[clap(after_help = EXAMPLE)]
pub struct RemoveAdmin {
    /// Name of the group. Can alternately refer to a group by ID. Indicate IDs by prefixing with 'id^' e.g. 'id^groupID'.
    #[clap(value_parser = util::group_identifier_from_string)]
    group: Either<GroupName, GroupId>,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(value_parser = clap::value_parser!(PathBuf), short, long)]
    keyfile: Option<PathBuf>,
    /// Remove admin permissions from the comma-separated list of user emails.
    #[clap(value_parser = util::try_from_email, short, long, use_value_delimiter = true, value_delimiter = ',', required = true)]
    users: Vec<UserId>,
}

impl util::GetKeyfile for RemoveAdmin {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn group_remove_admins(
    sdk: &BlockingIronOxide,
    RemoveAdmin { group, users, .. }: RemoveAdmin,
) -> Result<(), String> {
    let (groups_by_name, _) = get_group_maps(sdk);
    match convert_group_names_to_ids(&[group], &groups_by_name).first() {
        Some(group_id) => {
            let removed_admins = sdk.group_remove_admins(group_id, &users);
            let table = build_result_table(removed_admins);
            table.printstd();
        }
        None => {
            util::println_paint(Paint::red("Group does not exist.".to_string()));
        }
    }

    Ok(())
}

fn build_result_table(
    remove_admin_results: Result<GroupAccessEditResult, IronOxideErr>,
) -> prettytable::Table {
    let mut table = table!([Fbb->"User", Fbb->"Result"]);

    match remove_admin_results {
        Ok(results) => {
            for succeeded_id in results.succeeded() {
                table.add_row(row![
                    Fw -> succeeded_id.id(),
                    Fg -> format!("{}", "✔ Removed as admin"),
                ]);
            }
            for failure in results.failed() {
                table.add_row(row![
                    Fw -> failure.user().id(),
                    Fr -> failure.error(),
                ]);
            }
        }
        Err(e) => {
            table.add_row(
                row![Fw -> "ALL", Fr -> format!("Removing admins failed catastrophically: {}", e)],
            );
        }
    }

    table
}
